### 1. Importar las bibliotecas requeridas

In [1]:
import pandas as pd
%matplotlib inline

### 2. Importar la bd original y explorarla

In [2]:
df = pd.read_csv('bd titanic original.csv', encoding='latin-1')

In [3]:
df.head(5)

Id  Sobrevivió  Clase                                             Nombre  \
0   1           0      3                            Braund, Mr. Owen Harris   
1   2           1      1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2   3           1      3                             Heikkinen, Miss. Laina   
3   4           1      1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4   5           0      3                           Allen, Mr. William Henry   

     Sexo  Edad  HerEsp  PadHij            Ticket   Precio Cabina Embarcó  
0    male  22.0       1       0         A/5 21171   7.2500    NaN       S  
1  female  38.0       1       0          PC 17599  71.2833    C85       C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250    NaN       S  
3  female  35.0       1       0            113803  53.1000   C123       S  
4    male  35.0       0       0            373450   8.0500    NaN       S

In [4]:
print('Id: ', len(df.Id.unique()))
print('Sobrevivió: ', len(df.Sobrevivió.unique()))
print('Clase: ', len(df.Clase.unique()))
print('Sexo: ', len(df.Sexo.unique()))
print('HerEsp:', len(df.HerEsp.unique()))
print('PadHij: ', len(df.PadHij.unique()))
print('Ticket: ', len(df.Ticket.unique()))
print('Cabina: ', len(df.Cabina.unique()))
print('Embarcó: ', len(df.Embarcó.unique()))

Id:  891
Sobrevivió:  2
Clase:  3
Sexo:  2
HerEsp: 7
PadHij:  7
Ticket:  681
Cabina:  148
Embarcó:  4


**No vale la pena incluir la columna Id ya que es un ID de pasajeros que empieza en 1 y continua a un paso de +1.**

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Id          891 non-null    int64  
 1   Sobrevivió  891 non-null    int64  
 2   Clase       891 non-null    int64  
 3   Nombre      891 non-null    object 
 4   Sexo        891 non-null    object 
 5   Edad        714 non-null    float64
 6   HerEsp      891 non-null    int64  
 7   PadHij      891 non-null    int64  
 8   Ticket      891 non-null    object 
 9   Precio      891 non-null    float64
 10  Cabina      204 non-null    object 
 11  Embarcó     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**No vale la pena incluir la columna Cabina ya que posee un total de 148 valores únicos y solo tiene 204 filas con datos.**

In [6]:
df.describe()

Id  Sobrevivió       Clase        Edad      HerEsp      PadHij  \
count  891.000000  891.000000  891.000000  714.000000  891.000000  891.000000   
mean   446.000000    0.383838    2.308642   29.699118    0.523008    0.381594   
std    257.353842    0.486592    0.836071   14.526497    1.102743    0.806057   
min      1.000000    0.000000    1.000000    0.420000    0.000000    0.000000   
25%    223.500000    0.000000    2.000000   20.125000    0.000000    0.000000   
50%    446.000000    0.000000    3.000000   28.000000    0.000000    0.000000   
75%    668.500000    1.000000    3.000000   38.000000    1.000000    0.000000   
max    891.000000    1.000000    3.000000   80.000000    8.000000    6.000000   

           Precio  
count  891.000000  
mean    32.204208  
std     49.693429  
min      0.000000  
25%      7.910400  
50%     14.454200  
75%     31.000000  
max    512.329200

**¿Qué pasa con las celdas de la columna "Precio" que están en 0? ¿Cuáles son estas? ¿Son valores que hacen falta o realmente tuvieron un costo de $0?**

In [7]:
print(len(df[df['Precio'] == 0]))
df[df['Precio'] == 0]

15


Id  Sobrevivió  Clase                            Nombre  Sexo  Edad  \
179  180           0      3               Leonard, Mr. Lionel  male  36.0   
263  264           0      1             Harrison, Mr. William  male  40.0   
271  272           1      3      Tornquist, Mr. William Henry  male  25.0   
277  278           0      2       Parkes, Mr. Francis "Frank"  male   NaN   
302  303           0      3   Johnson, Mr. William Cahoone Jr  male  19.0   
413  414           0      2    Cunningham, Mr. Alfred Fleming  male   NaN   
466  467           0      2             Campbell, Mr. William  male   NaN   
481  482           0      2  Frost, Mr. Anthony Wood "Archie"  male   NaN   
597  598           0      3               Johnson, Mr. Alfred  male  49.0   
633  634           0      1     Parr, Mr. William Henry Marsh  male   NaN   
674  675           0      2        Watson, Mr. Ennis Hastings  male   NaN   
732  733           0      2              Knight, Mr. Robert J  male   NaN   
806  807           0      1            Andrews, Mr. Thomas Jr  male  39.0   
815  816           0      1                  Fry, Mr. Richard  male   NaN   
822  823           0      1   Reuchlin, Jonkheer. John George  male  38.0   

     HerEsp  PadHij  Ticket  Precio Cabina Embarcó  
179       0       0    LINE     0.0    NaN       S  
263       0       0  112059     0.0    B94       S  
271       0       0    LINE     0.0    NaN       S  
277       0       0  239853     0.0    NaN       S  
302       0       0    LINE     0.0    NaN       S  
413       0       0  239853     0.0    NaN       S  
466       0       0  239853     0.0    NaN       S  
481       0       0  239854     0.0    NaN       S  
597       0       0    LINE     0.0    NaN       S  
633       0       0  112052     0.0    NaN       S  
674       0       0  239856     0.0    NaN       S  
732       0       0  239855     0.0    NaN       S  
806       0       0  112050     0.0    A36       S  
815       0       0  112058     0.0   B102       S  
822       0       0   19972     0.0    NaN       S

In [8]:
precio_0 = df[df['Precio'] == 0]
print(precio_0['Clase'].value_counts())

2    6
1    5
3    4
Name: Clase, dtype: int64


**La columna "Embarcó" tiene dos celdas vacías o sin datos. ¿Cuáles son estas? ¿Por qué estos dos registros tienen el mismo número de ticket?**

In [9]:
df[df['Embarcó'].isnull()]

Id  Sobrevivió  Clase                                     Nombre  \
61    62           1      1                        Icard, Miss. Amelie   
829  830           1      1  Stone, Mrs. George Nelson (Martha Evelyn)   

       Sexo  Edad  HerEsp  PadHij  Ticket  Precio Cabina Embarcó  
61   female  38.0       0       0  113572    80.0    B28     NaN  
829  female  62.0       0       0  113572    80.0    B28     NaN

### 3. Lista de acciones a realizar en bd

*Acciones a realizar para limpiar y preprocesar después de exploración de la base de datos:*

1. Eliminar la columna Nombre, son 891 datos únicos que no brindarán ninguna tendencia.
2. Eliminar la columna Ticket, son 681 datos únicos que no generarán ninguna tendencia.
3. Eliminar la columna Cabina, pues el 77% de las celdas está en blanco o sin dato.
4. Considerar las celdas de Precio iguales a $0 a como que ese fue su precio real.
5. Imputar los valores en blanco de Embarcó, solo son 2 celdas, basado en la categoría más frecuente.
6. Crear columna con rangos de Edad para facilitar análisis y encontrar tendencias.
7. Crear columna que determine si la persona viajó sola o acompañada.
8. Crear una segunda base de datos, df2, que se enfoque solo en personas menor o igual a 18 años.
8. Crear columna que determine si los niños menor o igual a 18 años viajaban con hermanos o solos.
9. Crear columna que determine si los niños menor o igual a 18 años viajaban con padres o solos. 

### 4. Limpieza y preprocesamiento de datos

#### Acción 1, 2 y 3.

In [10]:
df = df.drop('Nombre', axis = 1)
df = df.drop('Ticket', axis = 1)
df = df.drop('Cabina', axis = 1)

#### Acción 5.

In [11]:
print(df['Embarcó'].value_counts())

S    644
C    168
Q     77
Name: Embarcó, dtype: int64


In [12]:
embarcó = df['Embarcó'].fillna(value = 'S')
df['Embarcó'] = embarcó

#### Acción 6.

In [13]:
df['RangosEdad'] = pd.cut(df['Edad'], bins=[0,13,18,29,49,90], right=True, labels=False) + 1

In [14]:
df['GruposEdad'] = df['RangosEdad'].map({1.0: 'Niño', 2.0: 'Adolescente', 3.0: 'Adulto jóven', 
                         4.0: 'Adulto', 5.0: 'Adulto mayor'})

In [15]:
df = df.drop('RangosEdad', axis = 1)

In [16]:
print(df['GruposEdad'].value_counts())

Adulto          256
Adulto jóven    245
Adulto mayor     74
Niño             71
Adolescente      68
Name: GruposEdad, dtype: int64


*Range 1 = 0-13 niño
..... Range 2 = 14-18 adolescente
..... Range 3 = 19-29 adulto jóven
..... Range 4 = 30-49 adulto
..... Range 5 = 50-90 adulto mayor*

#### Acción 7.

In [17]:
df['HerEsp + PadHij'] = df['HerEsp'] + df['PadHij']

In [18]:
df['RangosSolAcom'] = pd.cut(df['HerEsp + PadHij'], bins=[-1,0,20], right=True, labels=False) + 1

In [19]:
df['SolAcom'] = df['RangosSolAcom'].map({1.0: 'Solo', 2.0: 'Acompañado'})
df = df.drop('RangosSolAcom', axis = 1)

In [20]:
print(df['SolAcom'].value_counts())

Solo          537
Acompañado    354
Name: SolAcom, dtype: int64


#### Acción 8.

In [21]:
# df2 significa el dataframe 2 y esta solo contiene filas donde los pasajeros tienen menor o igual a 18 años.
df2 = df[df['Edad'] < 19]

#### Acción 9.

In [22]:
df2['RangosHerEsp'] = pd.cut(df2['HerEsp'], bins=[-1,0,2,15], right=True, labels=False) + 1
df2['GruposHerEsp'] = df2['RangosHerEsp'].map({1.0: 'Sin hermanos', 2.0: 'Entre 1-2', 3.0: 'Más de 2'})
df2 = df2.drop('RangosHerEsp', axis = 1)

<ipython-input-22-8045d93e1301>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['RangosHerEsp'] = pd.cut(df2['HerEsp'], bins=[-1,0,2,15], right=True, labels=False) + 1
<ipython-input-22-8045d93e1301>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['GruposHerEsp'] = df2['RangosHerEsp'].map({1.0: 'Sin hermanos', 2.0: 'Entre 1-2', 3.0: 'Más de 2'})


#### Acción 10.

In [23]:
df2['RangosPadHij'] = pd.cut(df2['PadHij'], bins=[-1,0,3], right=True, labels=False) + 1
df2['GruposPadHij'] = df2['RangosPadHij'].map({1.0: 'Sin padres', 2.0: 'Con 1-2 padres'})
df2 = df2.drop('RangosPadHij', axis = 1)

### 5. BDs listas para analizar y comer

**dataframe 1 or df1**

In [24]:
df.head(5)

Id  Sobrevivió  Clase    Sexo  Edad  HerEsp  PadHij   Precio Embarcó  \
0   1           0      3    male  22.0       1       0   7.2500       S   
1   2           1      1  female  38.0       1       0  71.2833       C   
2   3           1      3  female  26.0       0       0   7.9250       S   
3   4           1      1  female  35.0       1       0  53.1000       S   
4   5           0      3    male  35.0       0       0   8.0500       S   

     GruposEdad  HerEsp + PadHij     SolAcom  
0  Adulto jóven                1  Acompañado  
1        Adulto                1  Acompañado  
2  Adulto jóven                0        Solo  
3        Adulto                1  Acompañado  
4        Adulto                0        Solo

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               891 non-null    int64  
 1   Sobrevivió       891 non-null    int64  
 2   Clase            891 non-null    int64  
 3   Sexo             891 non-null    object 
 4   Edad             714 non-null    float64
 5   HerEsp           891 non-null    int64  
 6   PadHij           891 non-null    int64  
 7   Precio           891 non-null    float64
 8   Embarcó          891 non-null    object 
 9   GruposEdad       714 non-null    object 
 10  HerEsp + PadHij  891 non-null    int64  
 11  SolAcom          891 non-null    object 
dtypes: float64(2), int64(6), object(4)
memory usage: 83.7+ KB


In [26]:
df.to_csv('bd titanic mineros 1.csv', index = False, encoding = 'latin-1') 

**dataframe 2 or df2**

In [27]:
df2.head(5)

Id  Sobrevivió  Clase    Sexo  Edad  HerEsp  PadHij   Precio Embarcó  \
7    8           0      3    male   2.0       3       1  21.0750       S   
9   10           1      2  female  14.0       1       0  30.0708       C   
10  11           1      3  female   4.0       1       1  16.7000       S   
14  15           0      3  female  14.0       0       0   7.8542       S   
16  17           0      3    male   2.0       4       1  29.1250       Q   

     GruposEdad  HerEsp + PadHij     SolAcom  GruposHerEsp    GruposPadHij  
7          Niño                4  Acompañado      Más de 2  Con 1-2 padres  
9   Adolescente                1  Acompañado     Entre 1-2      Sin padres  
10         Niño                2  Acompañado     Entre 1-2  Con 1-2 padres  
14  Adolescente                0        Solo  Sin hermanos      Sin padres  
16         Niño                5  Acompañado      Más de 2  Con 1-2 padres

In [28]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 139 entries, 7 to 875
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               139 non-null    int64  
 1   Sobrevivió       139 non-null    int64  
 2   Clase            139 non-null    int64  
 3   Sexo             139 non-null    object 
 4   Edad             139 non-null    float64
 5   HerEsp           139 non-null    int64  
 6   PadHij           139 non-null    int64  
 7   Precio           139 non-null    float64
 8   Embarcó          139 non-null    object 
 9   GruposEdad       139 non-null    object 
 10  HerEsp + PadHij  139 non-null    int64  
 11  SolAcom          139 non-null    object 
 12  GruposHerEsp     139 non-null    object 
 13  GruposPadHij     139 non-null    object 
dtypes: float64(2), int64(6), object(6)
memory usage: 16.3+ KB


In [29]:
df2.to_csv('bd titanic mineros 2.csv', index = False, encoding = 'latin-1') 